File: 496-next-greater-element-i/next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        ans = []
        for i in range(len(nums1)):
            for j in range(len(nums2)):
                if nums2[j] == nums1[i]:
                    R = j
                    while R + 1 < len(nums2) and nums2[j] >= nums2[R]:
                        R += 1
                    ans.append(nums2[R] if nums2[R] > nums2[j] else -1)   
        return ans

File: 496-next-greater-element-i/next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        int n = nums1.size();
        unordered_map<int, int> mp;
        for (int i = 0; i < n; ++i) {
            mp[nums1[i]] = i;
        }

        vector<int> res(n, -1);
        stack<int> st;
        int idx;
        for (int i = 0; i < nums2.size(); ++i) {
            int num = nums2[i];
            while (!st.empty() && num > st.top()) {
                if (mp.find(st.top()) != mp.end()) {
                    idx = mp[st.top()];
                    res[idx] = num;
                }
                st.pop();
            }

            st.push(nums2[i]);
        }

        return res;

    }
};